File: src/vai_optimizer/pytorch_binding/pytorch_nndct/nn/src/cpu/bfp.cpp
/* Copyright 2023 Xilinx Inc.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*     http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/

#include <c10/core/DeviceType.h>

#include "../../include/bfp.h"
#include "../../../../../include/cpu/bfp.h"

torch::Tensor& to_bfp(const torch::Tensor& tensor,
                      int64_t bit_width,
                      int64_t block_size,
                      torch::Tensor& out) {
  if (tensor.device().type() == c10::DeviceType::CUDA) {
    LOG(FATAL) << "BFP operation only supports CPU, but got device type: "
               << tensor.device().type();
  }
  CheckInputForBFP(tensor, bit_width, block_size);

  //torch::Tensor result = at::empty_like(tensor);
  const float* input = tensor.data_ptr<float>();
  float* output = out.data_ptr<float>();

  LaunchBFPCPUKernel(input, output, tensor.numel(), bit_width, block_size);
  return out;
}

torch::Tensor& to_bfp_v2(const torch::Tensor& tensor,
                         int64_t bit_width,
                         int64_t block_size,
                         torch::Tensor& out) {
  if (tensor.device().type() == c10::DeviceType::CUDA) {
    LOG(FATAL) << "BFP operation only supports CPU, but got device type: "
               << tensor.device().type();
  }
  CheckInputForBFPV2(tensor, bit_width, block_size);

  //torch::Tensor result = at::empty_like(tensor);
  const float* input = tensor.data_ptr<float>();
  float* output = out.data_ptr<float>();

  LaunchBFPCPUKernelV2(input, output, tensor.numel(), bit_width, block_size);
  return out;
}

torch::Tensor& to_bfp_prime_shared(const torch::Tensor& tensor,
                                   int64_t bit_width,
                                   int64_t block_size,
                                   int64_t sub_block_size,
                                   int64_t sub_block_shift_bits,
                                   int64_t rounding_mode,
                                   torch::Tensor& out) {
  if (tensor.device().type() == c10::DeviceType::CUDA) {
    LOG(FATAL) << "BFP operation only supports CPU, but got device type: "
               << tensor.device().type();
  }
  CheckInputForBFPPrime(tensor, bit_width, block_size, sub_block_size);

  const float* input = tensor.data_ptr<float>();
  float* output = out.data_ptr<float>();

  LaunchBFPPrimeCPUKernel(input, output, tensor.numel(), bit_width,
        block_size, sub_block_size, sub_block_shift_bits, rounding_mode);

  return out;
}

//TORCH_LIBRARY_IMPL(vai, CPU, m) {
//  m.impl("to_bfp_v2", to_bfp_cpu)
//}

File: src/vai_quantizer/vai_q_pytorch/pytorch_binding/pytorch_nndct/nn/quantization/modules/bfp.py
# Copyright 2022 Xilinx Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

import torch
import torch.nn as nn
import torch.nn.functional as F

from pytorch_nndct.nn.quantization.ops import bfp_ops

class BFPQuantizer(nn.Module):

  def __init__(self,
               bitwidth,
               axis,
               round_mode='round_even',
               tile_size=8,
               is_prime=False,
               memory_format='channels_first',
               epsilon=torch.pow(torch.tensor(2.0), -23)):
    super(BFPQuantizer, self).__init__()

    self.bitwidth = bitwidth
    self.axis = axis
    self.round_mode = round_mode
    self.tile_size = tile_size
    self.epsilon = epsilon

    self._forward_fn = bfp_ops.quantize_to_bfp_prime if is_prime else bfp_ops.quantize_to_bfp

  def forward(self, input):
    return self._forward_fn(input, self.bitwidth, self.tile_size,
                            self.axis, self.round_mode, self.epsilon)

  def extra_repr(self):
    return 'bitwidth={}, tile_size={}, axis={}, round_mode={}, epsilon={}'.format(
        self.bitwidth, self.tile_size, self.axis, self.round_mode, self.epsilon)
